h/security/permits/engine.py

```diff
@@ -0,0 +1,69 @@
+from enum import Enum
+
+
+def predicate(requires=None):
+    def decorator(function):
+        function.requires = requires or []
+        return function
+
+    return decorator
+
+
+class PredicateBasedPermissions:
+    def __init__(self, mapping):
+        self.mapping = self._expand_clauses(mapping)
+
+    def permits(self, identity, context, permission):
+        if clauses := self.mapping.get(permission):
+            statement_cache = {}
```

Comment:
It could be a good idea to share this statement cache between evaluations on the same request. This could be a good candidate for: `RequestLocalCache`

https://docs.pylonsproject.org/projects/pyramid/en/latest/api/request.html#pyramid.request.RequestLocalCache

Which decorates a function to get a result which is cached per request. Which is very neat, and I can think of a few places that might be useful.

Reply:
Turns out this is probably a waste of time with our current predicate set etc. It's quicker to just do it again